### ISSUE TITLE
`libtheora` compiling errors for ARM (Hard Float)

### ISSUE THREAD

**Issue by silverkorn** [godotengine/godot#23506]
**Godot version:**
master (121cead38e5ea84ec22139df02eee56d822290b7)

**OS/device including version:**
Ubuntu 18.04.1 LTS through VirtualBox

**Issue description:**
Cross-compiling to `arm-linux-gnueabihf` fails when building `libtheora`.
It seems to be missing the `arm` directory from the original sources and the scons instructions. 

_Error print_:
```
[...]
[Initial build] Compiling ==> modules/svg/register_types.cpp
[Initial build] Compiling ==> modules/tga/image_loader_tga.cpp
[Initial build] Compiling ==> modules/tga/register_types.cpp
[Initial build] Compiling ==> thirdparty/libtheora/cpu.c
thirdparty/libtheora/cpu.c: In function 'oc_cpu_flags_get':
thirdparty/libtheora/cpu.c:42:3: error: impossible constraint in 'asm'
   __asm__ __volatile__( \
   ^
thirdparty/libtheora/cpu.c:157:3: note: in expansion of macro 'cpuid'
   cpuid(0,eax,ebx,ecx,edx);
   ^~~~~
thirdparty/libtheora/cpu.c:42:3: error: impossible constraint in 'asm'
   __asm__ __volatile__( \
   ^
thirdparty/libtheora/cpu.c:163:5: note: in expansion of macro 'cpuid'
     cpuid(1,eax,ebx,ecx,edx);
     ^~~~~
thirdparty/libtheora/cpu.c:42:3: error: impossible constraint in 'asm'
   __asm__ __volatile__( \
   ^
thirdparty/libtheora/cpu.c:171:5: note: in expansion of macro 'cpuid'
     cpuid(0x80000000,eax,ebx,ecx,edx);
     ^~~~~
thirdparty/libtheora/cpu.c:42:3: error: impossible constraint in 'asm'
   __asm__ __volatile__( \
   ^
thirdparty/libtheora/cpu.c:174:7: note: in expansion of macro 'cpuid'
       cpuid(0x80000001,eax,ebx,ecx,edx);
       ^~~~~
thirdparty/libtheora/cpu.c:42:3: error: impossible constraint in 'asm'
   __asm__ __volatile__( \
   ^
thirdparty/libtheora/cpu.c:178:5: note: in expansion of macro 'cpuid'
     cpuid(1,eax,ebx,ecx,edx);
     ^~~~~
thirdparty/libtheora/cpu.c:42:3: error: impossible constraint in 'asm'
   __asm__ __volatile__( \
   ^
thirdparty/libtheora/cpu.c:193:5: note: in expansion of macro 'cpuid'
     cpuid(1,eax,ebx,ecx,edx);
     ^~~~~
thirdparty/libtheora/cpu.c:42:3: error: impossible constraint in 'asm'
   __asm__ __volatile__( \
   ^
thirdparty/libtheora/cpu.c:206:7: note: in expansion of macro 'cpuid'
       cpuid(0x80000001,eax,ebx,ecx,edx);
       ^~~~~
{standard input}: Assembler messages:
{standard input}:254: Error: bad instruction `pushfl'
{standard input}:255: Error: bad instruction `pushfl'
{standard input}:256: Error: bad instruction `popl r3'
{standard input}:257: Error: bad instruction `movl r3,r2'
{standard input}:258: Error: bad instruction `xorl $0x200000,r3'
{standard input}:259: Error: bad instruction `pushl r3'
{standard input}:260: Error: bad instruction `popfl'
{standard input}:261: Error: bad instruction `pushfl'
{standard input}:262: Error: bad instruction `popl r3'
{standard input}:263: Error: bad instruction `popfl'
scons: *** [thirdparty/libtheora/cpu.x11.tools.arm.o] Error 1
scons: building terminated because of errors.
```

**Steps to reproduce:**
Using this Gist for cross-compiling ([godot-cross-compile.sh](https://gist.github.com/silverkorn/f0566819025d4a4882b82b2b27d20826)) or this command:
```bash
export CROSSFLAGS="-D__ARM_ARCH_8A__ -DGLES2_ENABLED -march=armv8-a -mtune=cortex-a53 -mfpu=neon-fp-armv8 -mfloat-abi=hard"

sudo scons CCFLAGS='${CROSSFLAGS}' CPPFLAGS='${CROSSFLAGS}' CC='arm-linux-gnueabihf-gcc' CXX='arm-linux-gnueabihf-g++' LINKFLAGS='-L/usr/arm-linux-gnueabihf/lib -L/usr/lib/arm-linux-gnueabihf' arch=arm builtin_freetype=yes platform=x11 -j 4
```

Thanks!

**Comment by bruvzg** [comment]
Are you sure you aren't mixing architectures? CPPFLAGS are for 64-bit ARMv8a and compiler/libs are for 32-bit ARMv7a `arm-linux-gnueabihf` (for 64-bit it should be `aarch64-linux-gnu`).

Building 121cead in Debian native ARM chroots works fine.

Command used for 32-bit / ARMv7a:
```
scons -j4 p=x11 tools=yes bits=32 target=debug use_static_cpp=yes use_llvm=yes builtin_freetype=yes builtin_libpng=yes builtin_zlib=yes extra_suffix=armv7 CXXFLAGS="-D__ARM_ARCH_7__ -D__ARM_ARCH_7A__ -march=armv7-a -mfpu=neon -mfloat-abi=hard"
```

Command used for 64-bit / ARMv8a:
```
scons -j4 p=x11 tools=yes bits=64 target=debug use_static_cpp=yes use_llvm=yes builtin_freetype=yes builtin_libpng=yes builtin_zlib=yes extra_suffix=armv8 CXXFLAGS="-D__ARM_ARCH_8A__ -march=armv8-a"
```

**Comment by bruvzg** [comment]
Also scons `arch` option is used only for iOS (added in https://github.com/godotengine/godot/commit/10c512e46d57034b2fd6f408e0fec37f17e972e5) it doesn't affect X11 build in any way.

**Comment by bruvzg** [comment]
It's not GCC/LLVM but scons build script issues:

https://github.com/godotengine/godot/blob/master/platform/x11/detect.py#L211
https://github.com/godotengine/godot/blob/master/modules/webm/libvpx/SCsub#L273

Scons is enabling architecture specific optimizations for `libtheora` and `libvpx` based on `platform.machine()` value.

Building inside QEMU chroot or on real ARM hardware works fine, cross-building doesn't since `platform.machine()` return 'x86_64'/'i686'.

**Comment by silverkorn** [comment]
Since it's a cross-compiling issue and it may breaks some part with the naming convention, I think we can stall it until the new output binary file naming will use the target architecture instead.

Let's lower the hassle for the intensive other bugs to release 3.1 :)

**Comment by akien-mga** [comment]
Thanks, that's a good point. Let's move to 3.2 milestone to reassess it once we've released 3.1, and likely include those changes in the aforementioned bigger refactor of binary file naming.

**Comment by akien-mga** [comment]
Related to #19319.

**Comment by Calinou** [comment]
Moving the milestone to 4.0 as #24030 was also moved to that milestone.

**Comment by akien-mga** [comment]
Presumably fixed by #55778. At least as of 3.6 and 4.2, we compile official Linux arm32 and arm64 builds including libtheora without issue.
